src/Compilers/VisualBasic/Test/IOperation/IOperation/IOperationTests_ILockStatement.vb

```diff
@@ -461,6 +497,92 @@ Block[B0] - Entry
     }
 }
 
+Block[B6] - Exit
+    Predecessors: [B2]
+    Statements (0)
+]]>.Value
+
+            VerifyFlowGraphAndDiagnosticsForTest(Of MethodBlockSyntax)(source, expectedFlowGraph, expectedDiagnostics)
+        End Sub
+
+        <Fact, CompilerTrait(CompilerFeature.IOperation, CompilerFeature.Dataflow)>
+        Public Sub LockFlow_LockObject()
+            Dim source = <![CDATA[
+Imports System.Threading
+
+Public Class C
+    Sub M(l As Lock) 'BIND:"Sub M"
+        SyncLock l
+        End SyncLock
+    End Sub
+End Class
+
+Namespace System.Threading
+    Public Class Lock
+    End Class
+End Namespace]]>.Value
+
+            Dim expectedDiagnostics = <![CDATA[
+BC37329: A value of type 'System.Threading.Lock' in SyncLock will use likely unintended monitor-based locking. Consider manually calling 'Enter' and 'Exit' methods in a Try/Finally block instead.
+        SyncLock l
+                 ~
+]]>.Value
+
+            Dim expectedFlowGraph = <![CDATA[
+Block[B0] - Entry
+    Statements (0)
+    Next (Regular) Block[B1]
+        Entering: {R1}
+.locals {R1}
+{
+    CaptureIds: [0]
+    Block[B1] - Block
+        Predecessors: [B0]
+        Statements (1)
+            IFlowCaptureOperation: 0 (OperationKind.FlowCapture, Type: null, IsInvalid, IsImplicit) (Syntax: 'l')
+              Value:
+                IConversionOperation (TryCast: False, Unchecked) (OperationKind.Conversion, Type: System.IDisposable, IsInvalid, IsImplicit) (Syntax: 'l')
+                  Conversion: CommonConversion (Exists: False, IsIdentity: False, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+                    (DelegateRelaxationLevelNone)
+                  Operand:
+                    IInvalidOperation (OperationKind.Invalid, Type: null, IsInvalid, IsImplicit) (Syntax: 'l')
```

Comment:
>IInvalidOperation [](http://example.com/codeflow?start=20&length=17)

I'm curious why we're not generating an `IInvocationOperation` that calls `Monitor.Enter()` here. From the binding code, it looks like we're reporting `ERR_LockTypeUnsupported` without marking the expression as invalid so it seems like `IOperation` and `ControlFlowGraphBuilder` would both treat this as a valid `SyncLock`. 

I'm not suggesting we *should* generate a valid `IOperation` tree here, but it's not clear to me why we're not currently.

Reply:
> From the binding code, it looks like we're reporting ```ERR_LockTypeUnsupported``` without marking the expression as invalid

First, there is no requirement to mark anything as invalid just because an error is reported. There is simply no contract like that.

Control flow graph builder creates invalid operation here because a helper is missing that it is supposed to call here.

Reply:
The original IOperation tree doesn't have corresponding node for it.

Reply:
That said, we do need to have a test for a scenario when all helpers are available